File: frontend/src/networking/SubConnection.ts
import Connection from "./Connection";
import ConnectionBase from "./ConnectionBase";
import { ConnectionAnswer, ConnectionOffer, ConnectionProposal } from "./typing";

/**
 * SubConnection class used by {@link Connection} to get streams of other users from the backend.
 * 
 * Not intended for use outside of {@link Connection}.
 * 
 * @see https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol for details about the connection protocol.
 * @extends ConnectionBase
 */
export default class SubConnection extends ConnectionBase<MediaStream | string> {
  readonly id: string;

  private connection: Connection;
  private stopped: boolean;

  /**
   * Initialize new SubConnection.
   * @param proposal ConnectionProposal received from the backend.
   * @param connection parent Connection, used to send data to the backend.
   * @param logging Whether logging should be enabled.
   * 
   * @see https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol for details about the connection protocol.
   */
  constructor(proposal: ConnectionProposal, connection: Connection, logging: boolean) {
    super(true, `SubConnection - ${proposal.id}`, logging);
    this.id = proposal.id;
    this.connection = connection;
    this.stopped = false;
    this._participantSummary = proposal.participant_summary;

    this.log("Initiating SubConnection");
  }

  /**
   * Send offer for this SubConnection to the backend.
   * 
   * @see https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol for details about the connection protocol.
   */
  public async sendOffer() {
    this.log("Generating & sending offer");
    const offer = await this.createOffer();
    const connectionOffer: ConnectionOffer = {
      id: this.id,
      offer: offer
    };
    this.connection.sendMessage("CONNECTION_OFFER", connectionOffer);
  }

  /** 
   * Handle `CONNECTION_ANSWER` message from the backend.
   * 
   * The answer is set as remote description for {@link pc}.
   */
  public async handleAnswer(answer: ConnectionAnswer) {
    // TODO Check if answer already handled
    await this.pc.setRemoteDescription(answer.answer);
  }

  /**
   * Stop the SubConnection.
   * 
   * Stop all transceivers associated with this SubConnection and its peer connection. 
   * 
   * Multiple calls to this functions are ignored.
   * @see https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol for details about the connection protocol.
   */
  public stop() {
    if (this.stopped) {
      return;
    }
    this.stopped = true;
    this.log("Stopping");

    // close transceivers
    this.pc.getTransceivers().forEach(function (transceiver) {
      if (transceiver.stop) {
        transceiver.stop();
      }
    });

    this.pc.close();
    this.emit("connectionClosed", this.id);
  }

  protected handleIceConnectionStateChange(): void {
    this.log(`IceConnectionState: ${this.pc.iceConnectionState}`);
    if (["disconnected", "closed", "failed"].includes(this.pc.iceConnectionState)) {
      this.stop();
    }
  }
}

File: backend/connection/sub_connection.py
import logging

from aiortc import RTCPeerConnection, MediaStreamTrack, RTCSessionDescription
from pyee.asyncio import AsyncIOEventEmitter

from connection.messages import (
    ConnectionAnswerDict,
    ConnectionProposalDict,
    RTCSessionDescriptionDict,
)
from session.data.participant.participant_summary import ParticipantSummaryDict


class SubConnection(AsyncIOEventEmitter):
    """SubConnection used to send additional stream to a client.

    Each SubConnection has an audio and video track -> one stream which they send to the
    client.  For communication, it used the datachannel of the parent
    module.connection.Connection.

    Extends AsyncIOEventEmitter, providing the following events:
    - `connection_closed` : str
        Emitted when this SubConnection closes.  Data is the ID of this SubConnection.
    """

    id: str
    _participant_summary: ParticipantSummaryDict | str | None
    _pc: RTCPeerConnection

    _audio_track: MediaStreamTrack
    _video_track: MediaStreamTrack

    _closed: bool
    _logger: logging.Logger

    def __init__(
        self,
        id: str,
        video_track: MediaStreamTrack,
        audio_track: MediaStreamTrack,
        participant_summary: ParticipantSummaryDict | str | None,
        log_name_suffix: str,
    ) -> None:
        """Initialize new SubConnection.

        Parameters
        ----------
        id : str
        video_track : aiortc.MediaStreamTrack
            Video track that will be sent in this SubConnection.
        audio_track : aiortc.MediaStreamTrack
            Audio track that will be sent in this SubConnection.
        participant_summary : None or custom_types.participant_summary.ParticipantSummaryDict
            Participant summary or ID send to the client with the initial offer.
        """
        super().__init__()
        self._logger = logging.getLogger(f"SubConnection-{log_name_suffix}")
        self.id = id
        self._audio_track = audio_track
        self._video_track = video_track
        self._closed = False
        self._participant_summary = participant_summary

        self._pc = RTCPeerConnection()
        self._pc.addTrack(video_track)
        self._pc.addTrack(audio_track)
        self._pc.on("connectionstatechange", self._on_connection_state_change)

        # Stop SubConnection if one of the tracks ends
        # audio_track.on("ended", self.stop)
        # video_track.on("ended", self.stop)

    @property
    def proposal(self):
        """Get a custom_types.connection.ConnectionProposalDict for this SubConnection.

        See Also
        --------
        Connection Protocol Wiki :
            https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol
        """
        return ConnectionProposalDict(
            id=self.id, participant_summary=self._participant_summary
        )

    async def stop(self) -> None:
        """Stop the SubConnection and its associated tracks.

        Emits a `connection_closed` event with the ID of this SubConnection.  When
        finished, it removes all event listeners from this Connection, as no more events
        will be emitted.
        """
        if self._closed:
            return
        self.emit("connection_closed", self.id)

        self._audio_track.stop()
        self._video_track.stop()

        self._logger.debug("Closing SubConnection")
        self._closed = True
        await self._pc.close()
        self.remove_all_listeners()

    async def handle_offer(self, offer: RTCSessionDescription) -> ConnectionAnswerDict:
        """Handle a `CONNECTION_OFFER` message for this SubConnection.

        Parameters
        ----------
        offer : aiortc.RTCSessionDescription

        Returns
        -------
        connection.messages.connection_answer_dict.ConnectionAnswerDict
            Answer that should be sent as a response to the offer.

        See Also
        --------
        Connection Protocol Wiki :
            https://github.com/TUMFARSynchorny/experimental-hub/wiki/Connection-Protocol
        """
        await self._pc.setRemoteDescription(offer)

        answer = await self._pc.createAnswer()
        await self._pc.setLocalDescription(answer)  # type: ignore

        answer_dict = RTCSessionDescriptionDict(
            sdp=self._pc.localDescription.sdp, type=self._pc.localDescription.type  # type: ignore
        )
        return ConnectionAnswerDict(id=self.id, answer=answer_dict)

    async def _on_connection_state_change(self):
        """Handle connection state change."""
        self._logger.debug(f"Peer Connection state change: {self._pc.connectionState}")
        if self._pc.connectionState in ["closed", "failed"]:
            await self.stop()
